### DISCUSSION TITLE
Singularity problems

### DISCUSSION THREAD

**Discussion by mflynn-lanl** [microbiomedata/nmdc-edge#258] [General]
I've had problems with two of the clusters I've created. For one, Singularity was not installed. For the second, Singularity was installed but `squashfuse` was not. I was able to install it manually

**Comment by mflynn-lanl** [comment]
I was able to install `squashfuse` on the login node, but not on the compute nodes

**Reply by mflynn-lanl** [reply]
i added the path to the `squashfuse` to PATH in /etc/profile.d/custom.sh. When I start an interactive slurm job, I see that the path to the squashfuse directory is in the PATH. However, when I run the Cromwell workflow, I get the same error: 
```
DEBUG   [U=60005,P=15715]  init()                        squashfuse mounting not enabled because: exec: "squashfuse": executable file not found in $PATH
```

If I launch the sbatch script manually, `sbatch cromwell_4f0bd364_int.sl`, I do not get the squashfuse error

**Comment by mflynn-lanl** [comment]
I created a second cluster with the same problem:
```
INFO:    squashfuse not found, will not be able to mount SIF
FATAL:   container creation failed: mount hook function failure: mount /proc/self/fd/11->/software/r9/apptainer/1.1.6/var/apptainer/mnt/session/rootfs error: while mounting image /proc/self/fd/11: squashfuse not found
```

**Reply by edwins** [reply]
Hi @mflynn-lanl ,

Could you tag my username, `edwins`, for conversations that need my attention? That would ensure I get notifications for future conversations.

As for singularity and other dependencies, I suggest creating a new image with all the tools you want into a base image (Alma9). You can simply launch as basic vm in any of the other JS2 UIs, and then create a "snapshot". By preinstalling your dependencies, this would eliminate the need in using one of the software stacks in the Magic Castle deployment wizard. Unfortunately, CACAO doesn't support creating individual snapshots  at this time. 

Here are links to instructions in creating snapshots using horizon, openstack cli, or exosphere:

https://docs.jetstream-cloud.org/ui/cli/snapshot-image/#snapshots-and-images
https://docs.jetstream-cloud.org/ui/horizon/manage/
https://docs.jetstream-cloud.org/getting-started/snapshots/

After your image is created, you should be able to select the image in the Boot image drop down.

I hope this helps.

Edwin

**Comment by mflynn-lanl** [comment]
Hi Edwin,
I've tried creating snapshots of my image in the past but I haven't been able to create a VM the deploys successfully
Mark
________________________________
From: Edwin Skidmore ***@***.***>
Sent: Friday, July 19, 2024 1:50 PM
To: microbiomedata/nmdc-edge ***@***.***>
Cc: Flynn, Mark C ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [microbiomedata/nmdc-edge] Singularity problems (Discussion #258)


Hi @mflynn-lanl<https://urldefense.com/v3/__https://github.com/mflynn-lanl__;!!Bt8fGhp8LhKGRg!GljecYWgIYgvnrdIpCRW6HLAofVzWZhWradEXTnFZbdGzwNjYy5_m_xiajthkM_46psb6tIDTlaqbnI1Nm65D5ly$> ,

Could you tag my username, edwins, for conversations that need my attention? That would ensure I get notifications for future conversations.

As for singularity and other dependencies, I suggest creating a new image with all the tools you want into a base image (Alma9). You can simply launch as basic vm in any of the other JS2 UIs, and then create a "snapshot". By preinstalling your dependencies, this would eliminate the need in using one of the software stacks in the Magic Castle deployment wizard. Unfortunately, CACAO doesn't support creating individual snapshots at this time.

Here are links to instructions in creating snapshots using horizon, openstack cli, or exosphere:

https://docs.jetstream-cloud.org/ui/cli/snapshot-image/#snapshots-and-images<https://urldefense.com/v3/__https://docs.jetstream-cloud.org/ui/cli/snapshot-image/*snapshots-and-images__;Iw!!Bt8fGhp8LhKGRg!GljecYWgIYgvnrdIpCRW6HLAofVzWZhWradEXTnFZbdGzwNjYy5_m_xiajthkM_46psb6tIDTlaqbnI1Nu5w6urt$>
https://docs.jetstream-cloud.org/ui/horizon/manage/<https://urldefense.com/v3/__https://docs.jetstream-cloud.org/ui/horizon/manage/__;!!Bt8fGhp8LhKGRg!GljecYWgIYgvnrdIpCRW6HLAofVzWZhWradEXTnFZbdGzwNjYy5_m_xiajthkM_46psb6tIDTlaqbnI1Nja8C_7k$>
https://docs.jetstream-cloud.org/getting-started/snapshots/<https://urldefense.com/v3/__https://docs.jetstream-cloud.org/getting-started/snapshots/__;!!Bt8fGhp8LhKGRg!GljecYWgIYgvnrdIpCRW6HLAofVzWZhWradEXTnFZbdGzwNjYy5_m_xiajthkM_46psb6tIDTlaqbnI1NmBBzDI-$>

After your image is created, you should be able to select the image in the Boot image drop down.

I hope this helps.

Edwin

**Reply by mflynn-lanl** [reply]
@edwins I'm creating a VM using Cacao `openstack-single-image` that will be used for the snapshot. Can I add `user` to that VM that will be used by the MC cluster? Do I need to set the UID and GID to what MC uses?

**Reply by edwins** [reply]
That is a great question. I suspect any system user will remain on that system once the snapshot is created. I would avoid any user or group (by name or id) that is created by magic castle as that might interfere with its operation.

**Reply by mflynn-lanl** [reply]
I'm worried that unless `user` is created by MC, that it won't have access to `/project`

**Reply by edwins** [reply]
As a clarification, you shouldn't add any users to your snapshot that might conflict with magic castle. Doing so, might prevent MC from working. In other words, I don't recommend adding a `user` account into snapshot or any user with the same uid.

**Comment by mflynn-lanl** [comment]
I've tried creating MC clusters using VM snapshots created from openstack, horizon and from exosphere, but all encountered an error during cluster creation

**Reply by edwins** [reply]
I can take a look. Just a couple of questions before I try to replicate the process:

Are you able to launch and login to a vm snapshot after creating it, using a simple launch (no magic castle)?

 Other than installing software, are you making any other changes on the vm?

**Comment by mflynn-lanl** [comment]
I can ssh to the VM created in exosphere, but not using Cacao simple launch or horizon
________________________________
From: Edwin Skidmore ***@***.***>
Sent: Monday, July 22, 2024 12:59 PM
To: microbiomedata/nmdc-edge ***@***.***>
Cc: Flynn, Mark C ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [microbiomedata/nmdc-edge] Singularity problems (Discussion #258)


I can take a look. Just a couple of questions before I try to replicate the process:

Are you able to launch and login to a vm snapshot after creating it, using a simple launch (no magic castle)?

Other than installing software, are you making any other changes on the vm?

**Reply by mflynn-lanl** [reply]
I created this VM using Horizon but I'm not able to ssh to it:
[https://js2.jetstream-cloud.org/project/instances/eb0e465e-ba3d-4a0e-9d7e-50139f2379f8/](url)

**Reply by mflynn-lanl** [reply]
I can ping the ip address, but I cannot ssh to it. I tried `ssh centos@public_ip_address`

**Reply by edwins** [reply]
When launching magic castle, I'm seeing connectivity errors, like:
```
2024-07-24T02:45:24.899Z [ERROR] connection error: dial tcp w.x.y.z:22: i/o timeout
2024-07-24T02:45:24.899Z [WARN]  retryable error: dial tcp w.x.y.z:22: i/o timeout
2024-07-24T02:45:25.289Z [ERROR] connection error: Error connecting to bastion: dial tcp w.x.y.z:22: connect: connection refused
2024-07-24T02:45:25.289Z [WARN]  retryable error: Error connecting to bastion: dial tcp w.x.y.z:22: connect: connection refused
2024-07-24T02:45:25.901Z [ERROR] connection error: dial tcp w.x.y.z:22: connect: connection refused
2024-07-24T02:45:25.901Z [WARN]  retryable error: dial tcp w.x.y.z:22: connect: connection refused
```
If I had to guess, it might be related to the networking of the instance, probably related to the snapshotting. Because of this, the configuration of the vm for mc does not happen. Looking for the cause or workaround.

**Reply by edwins** [reply]
@mflynn-lanl I have a fix for you. Here are the steps you should use when making the snapshot -- it doesn't matter where you launch or make the snapshot:

1. launch instance (you should pick the smallest vm you can)
2. install your software
3. add the following file, `/etc/cloud/cloud.cfg.d/10_js2_metadata.cfg`. Indentation is important (those are two spaces per indent).
```
datasource:
  OpenStack:
    max_wait: 600
    timeout: 30
    retries: 50
```
5. stop/shutoff the instance
6. snapshot the instance

I have tested the process and was able to launch a several Magic Castles without error. Don't forget to use Alma9 for now. 

If you're interested in the root cause, cloud-init is a standard facility that is used to help with the configuration of vms, including ssh keys, user accounts, etc, and cloud-init uses an OpenStack metadata endpoint to retrieve configuration information. It looks like the OpenStack metadata endpoint isn't reachable within the default timeout/retries for snapshotted images. There could be any number of reasons for this, but I suspect it's because a new instance must detect and reconfigure/recover from the networking configuration at the time of the snapshot was made. It could simply be dhcp-related. In any case, this config change will just increase the timeout/retry for cloud-init.

**Reply by mflynn-lanl** [reply]
I was able to create a MC cluster. However, when I tried running slurm commands I get `[user@login1 jar]$ sinfo
slurm_load_partitions: Unable to contact slurm controller (connect failure)`

**Reply by mflynn-lanl** [reply]
I ssh'd to the management node and tried to re-start `slurmctld` but got an error. This is from the slurmctld.log: 
```
[2024-07-24T20:42:23.958] accounting_storage/slurmdbd: init: Accounting storage SLURMDBD plugin loaded
[2024-07-24T20:42:23.958] error: slurm_persist_conn_open_without_init: failed to open persistent connection to host:mgmt1:6819: Connection refused
[2024-07-24T20:42:23.958] error: Sending PersistInit msg: Connection refused
[2024-07-24T20:42:23.958] accounting_storage/slurmdbd: clusteracct_storage_p_register_ctld: Registering slurmctld at port 6817 with slurmdbd
[2024-07-24T20:42:23.958] error: Sending PersistInit msg: Connection refused
[2024-07-24T20:42:23.958] error: Association database appears down, reading from state file.
[2024-07-24T20:42:23.958] debug:  create_mmap_buf: Failed to open file `/var/spool/slurm/last_tres`, No such file or directory
[2024-07-24T20:42:23.958] debug:  create_mmap_buf: Failed to open file `/var/spool/slurm/assoc_mgr_state`, No such file or directory
[2024-07-24T20:42:23.958] error: Unable to get any information from the state file
[2024-07-24T20:42:23.958] fatal: slurmdbd and/or database must be up at slurmctld start time
[2024-07-24T20:42:23.959] error: Sending PersistInit msg: Connection refused
[2024-07-24T20:42:23.959] slurmscriptd: debug:  _slurmscriptd_mainloop: finished
```

**Reply by edwins** [reply]
I'll try to reproduce from my allocation.

**Reply by mflynn-lanl** [reply]
```
[mcflynn@mgmt1 ~]$ mysql -uslurm -p slurm_acct_db --host mgmt1
Enter password:
ERROR 1130 (HY000): Host 'mgmt1.int.open-snap-cluster.mcb180107.jetstream-cloud.org' is not allowed to connect to this MySQL server
```

**Reply by edwins** [reply]
I can't reproduce that error with a snapshot I created and slurm seems to work for me. Would you be able to share your snapshot with me?

Here are instructions: https://docs.jetstream-cloud.org/ui/cli/snapshot-image/#sharing-an-image

I can't seem to find instructions on image sharing using Horizon or even Exosphere. If you're ok using the openstack cli, the project to share with is `TRA220028`.

Alternatively, you're welcome to  me to your allocation for testing purposes.

**Comment by mflynn-lanl** [comment]
Hi Edwin,
I added you to our allocation.
Mark
________________________________
From: Edwin Skidmore ***@***.***>
Sent: Wednesday, July 24, 2024 5:57 PM
To: microbiomedata/nmdc-edge ***@***.***>
Cc: Flynn, Mark C ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [microbiomedata/nmdc-edge] Singularity problems (Discussion #258)


I can't reproduce that error with a snapshot I created and slurm seems to work. Would you be able to share your snapshot with me?

Here are instructions: https://docs.jetstream-cloud.org/ui/cli/snapshot-image/#sharing-an-image<https://urldefense.com/v3/__https://docs.jetstream-cloud.org/ui/cli/snapshot-image/*sharing-an-image__;Iw!!Bt8fGhp8LhKGRg!CFWLsz6UknuQXX6QluEi7uL-OOGG-Dbd1HOBFNuRbyx5VM3URq5ImIB8Ejxn5cKkvlJU0H1HzdxBFo69rx4VmotU$>

I can't seem to find instructions on image sharing using Horizon or even Exosphere. If you're find using the cli, the project to share with is TRA220028.

Alternatively, you're welcome to me to your allocation for testing purposes.

**Reply by edwins** [reply]
Which snapshot should I be testing?

**Reply by mflynn-lanl** [reply]
https://js2.jetstream-cloud.org/ngdetails/OS::Glance::Image/6a69b206-df76-4220-9eda-d87a0fe909c0

**Reply by mflynn-lanl** [reply]
I created a new MC cluster from a snapshot from a VM where I did not install anything. I just added that file `/etc/cloud/cloud.cfg.d/10_js2_metadata.cfg` and for this cluster, slurm commands worked. So it looks like something happens when I install the software on the VM

**Reply by mflynn-lanl** [reply]
I'm trying again with progressively more packages installed, first Java, then mysql

**Reply by mflynn-lanl** [reply]
If I install everything but MySQL, slurm works. Somehow installing MySQL messes up slurm

**Reply by edwins** [reply]
oops, it looks like my suspicions were correct (see the message I just posted below).

**Comment by mflynn-lanl** [comment]
```
sudo dnf install java-11-openjdk java-11-openjdk-devel -y && \
sudo dnf update -y && sudo dnf install mysql-server epel-release apptainer -y && sudo systemctl enable mysqld && sudo systemctl start mysqld && sudo mysql_secure_installation
mysql -uroot -p -e "create database cromwell;CREATE USER 'cromwell'@'localhost' IDENTIFIED BY '******';GRANT ALL PRIVILEGES ON cromwell.* TO 'cromwell'@'localhost' WITH GRANT OPTION;FLUSH PRIVILEGES;"
sudo bash -c  "echo 'export APPTAINER_CACHEDIR=/project/apptainer/cache' > /etc/profile.d/custom.sh"
```

**Reply by edwins** [reply]
I wouldn't be surprised if mysql was the issue, given that magic castle uses that as a dependency. 

When I have time, I'll see if magic castle supports configuring different images for each of the node types e.g. your snapshot for the worker nodes while using a base AlmaLinux9 for the mgmt node. If so, I can expose that on the CACAO interface.

**Reply by mflynn-lanl** [reply]
I only need MySQL running on the login node or I could even run it on a separate VM that I can connect to through the network.

**Reply by mflynn-lanl** [reply]
I created a new vm and installed mysql. I added the mysql vm to the same security group as my new MC cluster. I'm not sure if it's necessary, but I added the public key from the login node to the mysql vm. When I connect to mysql running on the new VM, for 'host' should I use the private IP address of the VM running mysql? For example `mysql -ucromwell -p -h 10.0.61.83 cromwell`

**Reply by edwins** [reply]
You should be able to use the private ip. It would be similar to how you exposed the webapp to the login node, in terms of access.

**Reply by mflynn-lanl** [reply]
Hmmm... Do you happen to have any experience with MySQL configuration? I'm not able to connect from the login node. 
`ERROR 2003 (HY000): Can't connect to MySQL server on '10.0.61.83:3306' (113)`

**Reply by mflynn-lanl** [reply]
I can ssh from the login node to the mysql VM. Does the firewall permit connections on port 3306?

**Reply by mflynn-lanl** [reply]
I created a user `'cromwell'@'10.0.61.83'` in the MySQL database and added `bind-address=10.0.61.83` to `/etc/my.cnf.d/mysql-server.cnf`
I also added a rule to the security group `os-debug-1-secgroup` using Horizon:
`Ingress | IPv4 | TCP | 3306 (MYSQL) | 0.0.0.0/0`

**Reply by mflynn-lanl** [reply]
I'm still not able to connect to mysql from the login node:
```
[mcflynn@login1 ~]$ nc -z -v -w 5 10.0.61.83 3306
Ncat: Version 7.92 ( https://nmap.org/ncat )
Ncat: No route to host.
```

**Reply by edwins** [reply]
Sorry, I misinformed you earlier. You'll need to ensure that the firewall/security group attached to the mysql host has tcp/3306 opened i.e. from the login node to the mysql host. You can probably just open it up to the subnet.

**Reply by mflynn-lanl** [reply]
I needed to restart the mysql VM and now it works!

**Reply by mflynn-lanl** [reply]
Cromwell can also connect to MySQL running on the VM so you don't need to do anything so that the login node and management nodes have different software installed during cluster creation

**Reply by edwins** [reply]
cool, i'm glad you got that working

**Reply by mflynn-lanl** [reply]
I'm no longer able to ssh to MC cluster login node for any of the clusters I created today, even though I could ssh to ones I created yesterday. For example: deployment-cqkq0j9ogabsfg6erl2g
https://cacao.jetstream-cloud.org/deployments/deployment-cqkq0j9ogabsfg6erl2g

**Comment by mflynn-lanl** [comment]
```
(base) mflynn@pn2294255 MIGDAL_simulator % ssh -vvv mcflynn@149.165.154.115
OpenSSH_9.6p1, LibreSSL 3.3.6
debug1: Reading configuration data /Users/mflynn/.ssh/config
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 21: include /etc/ssh/ssh_config.d/* matched no files
debug1: /etc/ssh/ssh_config line 54: Applying options for *
debug2: resolve_canonicalize: hostname 149.165.154.115 is address
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts' -> '/Users/mflynn/.ssh/known_hosts'
debug3: expanded UserKnownHostsFile '~/.ssh/known_hosts2' -> '/Users/mflynn/.ssh/known_hosts2'
debug1: Authenticator provider $SSH_SK_PROVIDER did not resolve; disabling
debug3: channel_clear_timeouts: clearing
debug3: ssh_connect_direct: entering
debug1: Connecting to 149.165.154.115 [149.165.154.115] port 22.
debug3: set_sock_tos: set socket 3 IP_TOS 0x48
```

**Reply by mflynn-lanl** [reply]
I think the problem was the length of the name of the cluster. I tried with a shorter name and now I can ssh to the MC cluster
